noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr

```diff
@@ -39,53 +39,60 @@ pub(crate) comptime fn transform_private(f: FunctionDefinition) {
     let (args_serialization, _, serialized_args_name) =
         derive_serialization_quotes(original_params, false);
 
-    let context_creation = quote {
-        $args_serialization
-        let args_hash = dep::aztec::hash::hash_args_array($serialized_args_name);
-        let mut context = dep::aztec::context::private_context::PrivateContext::new(inputs, args_hash);
+    let storage_init = if module_has_storage {
+        // Contract has Storage defined so we initialize it.
+        quote {
+            let storage = Storage::init(&mut context);
+        }
+    } else {
+        // Contract does not have Storage defined, so we initialize placeholder storage. ContractSelf requires
+        // a storage struct in its constructor. Using an Option type there would lead to worse developer experience
+        // and higher constraint counts so we use this placeholder struct instead.
+        quote {
+            let storage = aztec::placeholder_storage::PlaceholderStorage {};
+        }
+    };
+
+    let contract_self_creation = quote {
+        let mut self = {
+            $args_serialization
+            let args_hash = aztec::hash::hash_args_array($serialized_args_name);
+            let mut context = aztec::context::private_context::PrivateContext::new(inputs, args_hash);
+            $storage_init
+            aztec::contract_self::ContractSelf::new_private(&mut context, storage)
+        };
```

Comment:
Good thinking with putting this inside a block to avoid polluting the scope. We want to add `#[allow(unused_variables)]` to `self` though, like we did for `context`.

Reply:
Good point: https://github.com/AztecProtocol/aztec-packages/pull/17884/commits/759bd22ca23614987ee58113c96b51f666b826eb